eras/babbage/test-suite/src/Test/Cardano/Ledger/Babbage/Serialisation/Generators.hs

```diff
@@ -0,0 +1,348 @@
+import Cardano.Ledger.Babbage (BabbageEra)
+import Cardano.Ledger.Babbage.PParams
+import Cardano.Ledger.Babbage.Tx
+import Cardano.Ledger.Babbage.TxBody
+  ( BabbageBody,
+    Datum (..),
+    TxBody (..),
+    TxOut (..),
+  )
+import Cardano.Ledger.CompactAddress (compactAddr)
+import Cardano.Ledger.Compactible (toCompact)
+import qualified Cardano.Ledger.Core as Core
+import Cardano.Ledger.Era (Crypto, Era, ValidateScript (..))
+import Cardano.Ledger.Hashes (ScriptHash)
+import Cardano.Ledger.Shelley.Constraints (UsesScript, UsesValue)
+import Control.Applicative (liftA2)
+import Data.Int (Int64)
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe (fromJust, fromMaybe)
+import qualified Data.Set as Set
+import Data.Text (pack)
+import Numeric.Natural (Natural)
+import Plutus.V1.Ledger.Api (defaultCostModelParams)
+import qualified PlutusTx as Plutus
+import Test.Cardano.Ledger.Alonzo.Scripts (alwaysFails, alwaysSucceeds)
+import Test.Cardano.Ledger.Shelley.ConcreteCryptoTypes (Mock)
+import Test.Cardano.Ledger.Shelley.Serialisation.EraIndepGenerators ()
+import Test.Cardano.Ledger.ShelleyMA.Serialisation.Generators (genMintValues)
+import Test.QuickCheck
+
+instance
+  ( Era era,
+    UsesValue era,
+    Mock (Crypto era),
+    Arbitrary (Core.Value era),
+    Arbitrary (Core.Script era)
+  ) =>
+  Arbitrary (TxOut era)
+  where
+  arbitrary =
+    TxOut
+      <$> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+
+instance
+  ( UsesValue era,
+    Mock (Crypto era),
+    BabbageBody (era),
+    Arbitrary (Core.Value era),
+    Arbitrary (Core.Script era)
+  ) =>
+  Arbitrary (TxBody era)
+  where
+  arbitrary =
+    TxBody
+      <$> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> genMintValues
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+
+instance Arbitrary (Data era) where
+  arbitrary = Data <$> arbitrary
+
+genPair :: Gen a -> Gen b -> Gen (a, b)
+genPair = liftA2 (,)
+
+instance Arbitrary Plutus.Data where
+  arbitrary = resize 5 (sized gendata)
+    where
+      gendata n
+        | n > 0 =
+          oneof
+            [ (Plutus.I <$> arbitrary),
+              (Plutus.B <$> arbitrary),
+              (Plutus.Map <$> listOf (genPair (gendata (n `div` 2)) (gendata (n `div` 2)))),
+              ( Plutus.Constr <$> fmap fromIntegral (arbitrary :: Gen Natural)
+                  <*> listOf (gendata (n `div` 2))
+              ),
+              (Plutus.List <$> listOf (gendata (n `div` 2)))
+            ]
+      gendata _ = oneof [Plutus.I <$> arbitrary, Plutus.B <$> arbitrary]
+
+instance
+  ( UsesScript era,
+    Ord (Core.Script era),
+    Core.Script era ~ Script era,
+    Arbitrary (Core.Script era)
+  ) =>
+  Arbitrary (AuxiliaryData era)
```

Comment:
reuse alonzo? I'll stop making this comment now, but I think it applies to a lot of the generators below as well.